oauth/authcode.go

```diff
@@ -0,0 +1,183 @@
+	"net/url"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"context"
+
+	"github.com/danielgtaylor/openapi-cli-generator/cli"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/viper"
+	"golang.org/x/oauth2"
+	gcontext "gopkg.in/h2non/gentleman.v2/context"
+)
+
+// open opens the specified URL in the default browser regardless of OS.
+func open(url string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default: // "linux", "freebsd", "openbsd", "netbsd"
+		cmd = "xdg-open"
+	}
+	args = append(args, url)
+	return exec.Command(cmd, args...).Start()
+}
+
+// authHandler is an HTTP handler that takes a channel and sends the `code`
+// query param when it gets a request.
+type authHandler struct {
+	c chan string
+}
+
+func (h authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.c <- r.URL.Query().Get("code")
+	w.Header().Set("Content-Type", "text/html")
+	w.Write([]byte("<html><body><p>Login successful. Please return to the terminal. You may now close this window.</p></body></html>"))
+}
+
+// AuthorizationCodeTokenSource with PKCE as described in:
+// https://www.oauth.com/oauth2-servers/pkce/
+// This works by running a local HTTP server on port 8484 and then having the
+// user log in through a web browser, which redirects to the local server with
+// an authorization code. That code is then used to make another HTTP request
+// to fetch an auth token (and refresh token). That token is then in turn
+// used to make requests against the API.
+type AuthorizationCodeTokenSource struct {
+	ClientID       string
+	AuthorizeURL   string
+	TokenURL       string
+	EndpointParams *url.Values
+	Scopes         []string
+}
+
+// Token generates a new token using an authorization code.
+func (ac *AuthorizationCodeTokenSource) Token() (*oauth2.Token, error) {
+	// Generate a random code verifier string
+	verifierBytes := make([]byte, 32)
+	if _, err := rand.Read(verifierBytes); err != nil {
+		return nil, err
+	}
+
+	verifier := base64.RawURLEncoding.EncodeToString(verifierBytes)
+
+	// Generate a code challenge. Only the challenge is sent when requesting a
+	// code which allows us to keep it secret for now.
+	shaBytes := sha256.Sum256([]byte(verifier))
+	challenge := base64.RawURLEncoding.EncodeToString(shaBytes[:])
+
+	// Generate a URL with the challenge to have the user log in.
+	url := fmt.Sprintf("%s?response_type=code&code_challenge=%s&code_challenge_method=S256&client_id=%s&redirect_uri=http://localhost:8484/&scope=%s", ac.AuthorizeURL, challenge, ac.ClientID, strings.Join(ac.Scopes, `%20`))
+
+	if len(*ac.EndpointParams) > 0 {
+		url += "&" + ac.EndpointParams.Encode()
+	}
+
+	// Run server before opening the user's browser so we are ready for any redirect.
+	codeChan := make(chan string)
+	handler := authHandler{
+		c: codeChan,
+	}
+
+	s := &http.Server{
+		Addr:           ":8484",
+		Handler:        handler,
```

Comment:
Just curious due to my lack of go knowledge. Adding the handler to the server means that no matter which browser the user uses, we should detect the redirect with the returned code, correct?

Reply:
It's setting a default handler function for the HTTP server, regardless of route it will always get called. In the handler we pull out the returned code and send it into the channel.

This doesn't have anything to do with the browser, so yes it is browser-agnostic. Regardless of which browser you use, it will redirect to our server on `:8484` and we will get the code, shut down the server, request a token, cache it, and then run the requested API command and print the response.